src/event/utils.ts

```diff
@@ -25,10 +36,34 @@ export function defineEventHandler<
 >;
 export function defineEventHandler<
   Request extends EventHandlerRequest = EventHandlerRequest,
-  Response = EventHandlerResponse,
->(handler: EventHandler<Request, Response>): EventHandler<Request, Response> {
-  handler.__is_handler__ = true;
-  return handler;
+  Response = any,
+  Validator extends
+    ValidateFunction<EventHandlerRequest> = ValidateFunction<EventHandlerRequest>,
+  ValidatedRequest extends
+    EventHandlerRequest = Validator extends ValidateFunction<infer T>
+    ? T
+    : EventHandlerRequest,
+>(
+  handler: EventHandlerInput<Request, Response, Validator, ValidatedRequest>,
+): EventHandler<Request, Response> {
+  if (typeof handler === "function") {
+    return Object.assign(handler, { __is_handler__: true });
+  }
+  const wrapper: EventHandler<Request, any> = async (event) => {
+    if (handler.validate) {
+      await validateData(event, handler.validate);
```

Comment:
Let's not include validate in first iteration. I am worried (even if small, this adds to the bundle when not validation is needed). We can rethinking about adding more proper in next iterations 👍🏼

Reply:
Makes sense! Mainly I wanted to confirm that the types would work with inferring typed request _from_ a provided validator function.